extension/observer/hostobserver/extension.go

```diff
@@ -0,0 +1,252 @@
+		if c.Pid == 0 {
+			cd := collectConnectionDetails(&c)
+			id := observer.EndpointID(
+				fmt.Sprintf(
+					"(%s)%s-%d-%s", e.observerName, c.Laddr.IP, cd.port, cd.transport,
+				),
+			)
+
+			endpoints = append(endpoints, observer.Endpoint{
+				ID:     id,
+				Target: cd.target,
+				Details: observer.HostPort{
+					Port:      cd.port,
+					Transport: cd.transport,
+					// TODO: Move this field to observer.Endpoint and
+					// update receiver_creator to filter IPv4/IPv6.
+					IsIPv6: cd.isIPv6,
+				},
+			})
+			continue
+		}
+
+		connsByPID[c.Pid] = append(connsByPID[c.Pid], &c)
+	}
+
+	for pid, conns := range connsByPID {
+		proc, err := process.NewProcess(pid)
+		if err != nil {
+			e.logger.Warn("Could not examine process (it might have terminated already)")
+			continue
+		}
+
+		pd, err := collectProcessDetails(proc)
+		if err != nil {
+			e.logger.Error("failed collecting process details (skipping)",
+				zap.Int32("pid", pid), zap.Error(err),
+			)
+			continue
+		}
+
+		for _, c := range conns {
+			cd := collectConnectionDetails(c)
+
+			id := observer.EndpointID(
+				fmt.Sprintf(
+					"(%s)%s-%d-%s-%d",
+					e.observerName, c.Laddr.IP, cd.port, cd.transport, pid,
+				),
+			)
+
+			e := observer.Endpoint{
+				ID:     id,
+				Target: cd.target,
+				Details: observer.HostPort{
+					Name:      pd.name,
+					Command:   pd.args,
+					Port:      cd.port,
+					Transport: cd.transport,
+					// TODO: Move this field to observer.Endpoint and
+					// update receiver_creator to filter IPv4/IPv6.
+					IsIPv6: cd.isIPv6,
+				},
+			}
+			endpoints = append(endpoints, e)
+		}
+	}
+
+	return endpoints
+}
+
+type connectionDetails struct {
+	ip        string
+	isIPv6    bool
+	port      uint16
+	target    string
+	transport observer.Transport
+}
+
+func collectConnectionDetails(c *net.ConnectionStat) connectionDetails {
+	ip := c.Laddr.IP
+	// An IP addr of 0.0.0.0 means it listens on all interfaces, including
+	// localhost, so use that since we can't actually connect to 0.0.0.0.
+	if ip == "0.0.0.0" {
+		ip = "127.0.0.1"
+	}
+
+	isIPv6 := false
+
+	if c.Family == syscall.AF_INET6 {
+		ip = "[" + ip + "]"
+		isIPv6 = true
+	}
+
+	port := uint16(c.Laddr.Port)
+	target := fmt.Sprintf("%s:%d", ip, port)
+	protocol := portTypeToProtocol(c.Type)
+
+	return connectionDetails{
+		ip:        ip,
+		isIPv6:    isIPv6,
+		port:      port,
+		target:    target,
+		transport: protocol,
+	}
+}
+
+type processDetails struct {
+	name string
+	args string
+}
+
+func collectProcessDetails(proc *process.Process) (*processDetails, error) {
+	name, err := getProcessName(proc)
+	if err != nil {
+		return nil, fmt.Errorf("could not get process name: %v", err)
+	}
+
+	args, err := getProcessArgs(proc)
+	if err != nil {
+		return nil, fmt.Errorf("could not get process args: %v", err)
+	}
+
+	return &processDetails{
+		name: name,
+		args: args,
+	}, nil
+}
+
+var getProcessName = func(proc *process.Process) (string, error) {
```

Comment:
Maybe in followup PR you could try refactoring this into a struct and/or interface that can be injected by a mock implementation for tests. Should be safer than trying to modify package level variables in test. Can talk about it more offline if it's not clear what I mean.

Reply:
I think understand what you mean. Will make a subsequent PR to refactor.